src/xrpld/app/tx/detail/AMMClawback.cpp

```diff
@@ -0,0 +1,283 @@
+        JLOG(ctx.j.trace())
+            << "AMMClawback: holder cannot be the same as issuer.";
+        return temMALFORMED;
+    }
+
+    std::optional<STAmount> const clawAmount = ctx.tx[~sfAmount];
+    auto const asset = ctx.tx[sfAsset];
+
+    if (isXRP(asset))
+        return temMALFORMED;
+
+    if (asset.account != issuer)
+    {
+        JLOG(ctx.j.trace()) << "AMMClawback: Asset's account does not "
+                               "match Account field.";
+        return temBAD_ASSET_ISSUER;
+    }
+
+    if (clawAmount && clawAmount->issue() != asset)
+    {
+        JLOG(ctx.j.trace()) << "AMMClawback: Amount's issuer/currency subfield "
+                               "does not match Asset field";
+        return temBAD_ASSET_AMOUNT;
+    }
+
+    if (clawAmount && *clawAmount <= beast::zero)
+        return temBAD_AMOUNT;
+
+    return preflight2(ctx);
+}
+
+TER
+AMMClawback::preclaim(PreclaimContext const& ctx)
+{
+    auto const asset = ctx.tx[sfAsset];
+    auto const asset2 = ctx.tx[sfAsset2];
+    auto const sleIssuer = ctx.view.read(keylet::account(ctx.tx[sfAccount]));
+    if (!sleIssuer)
+        return terNO_ACCOUNT;  // LCOV_EXCL_LINE
+
+    auto const ammSle = ctx.view.read(keylet::amm(asset, asset2));
+    if (!ammSle)
+    {
+        JLOG(ctx.j.debug()) << "AMM Clawback: Invalid asset pair.";
+        return terNO_AMM;
+    }
+
+    std::uint32_t const issuerFlagsIn = sleIssuer->getFieldU32(sfFlags);
+
+    // If AllowTrustLineClawback is not set or NoFreeze is set, return no
+    // permission
+    if (!(issuerFlagsIn & lsfAllowTrustLineClawback) ||
+        (issuerFlagsIn & lsfNoFreeze))
+        return tecNO_PERMISSION;
+
+    auto const flags = ctx.tx.getFlags();
+    if (flags & tfClawTwoAssets && asset.account != asset2.account)
+    {
+        JLOG(ctx.j.trace())
+            << "AMMClawback: tfClawTwoAssets can only be enabled when two "
+               "assets in the AMM pool are both issued by the issuer";
+        return tecNO_PERMISSION;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+AMMClawback::doApply()
+{
+    Sandbox sb(&ctx_.view());
+
+    auto const ter = applyGuts(sb);
+    if (ter == tesSUCCESS)
+        sb.apply(ctx_.rawView());
+
+    return ter;
+}
+
+TER
+AMMClawback::applyGuts(Sandbox& sb)
+{
+    std::optional<STAmount> const clawAmount = ctx_.tx[~sfAmount];
+    AccountID const issuer = ctx_.tx[sfAccount];
+    AccountID const holder = ctx_.tx[sfHolder];
+    Issue const asset = ctx_.tx[sfAsset];
+    Issue const asset2 = ctx_.tx[sfAsset2];
+
+    auto ammSle = sb.peek(keylet::amm(asset, asset2));
+    if (!ammSle)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const ammAccount = (*ammSle)[sfAccount];
+    auto const accountSle = sb.read(keylet::account(ammAccount));
+    if (!accountSle)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const expected = ammHolds(
+        sb,
+        *ammSle,
+        asset,
+        asset2,
+        FreezeHandling::fhZERO_IF_FROZEN,
```

Comment:
I think there will be an issue here if the trustline is frozen. Wont the AMM return 0? Add tests for global and line freeze to confirm.

Reply:
Yes. Changed to fhIGNORE_FREEZE in recent commit. And also made related changes and added unit tests for global and individual freeze.